File bundle prefetching

vvolkl opened the pull request:
# File bundle prefetching

Introduce *file bundles*: a way to speedup application startup by declaring
 that opening one file should trigger prefetching of a set of related files.

A bundle is described by a `.cvmfsbundle-<filename>` specification file
placed next to its trigger `<filename>` in the same directory. The spec is a
small versioned JSON document listing the files to prefetch:

```json
{
  "name": "CVMFS_BUNDLE",
  "version": "1.0.0",
  "encoding": "UTF-8",
  "dependencies": [
    "/absolute/path/from/repo/root/libfoo.so",
    "/absolute/path/from/repo/root/data.bin"
  ]
}
```

On the publisher side, the sync mediator picks up these specification files
and links each bundle to its trigger file (exposed via the `bundle_trigger`
magic xattr). When a client then `open()`s the trigger file, a dedicated
fetcher thread pool prefetches and decompresses the listed dependencies in
parallel.

## User-facing changes

- **Client option `CVMFS_PREFETCH_FILEBUNDLES`** (boolean, off by default)
   to enable bundle prefetching on all repositories. 
- **`.cvmfsbundle-<filename>` specification files** are recognized by the
  publisher; `cvmfs_server` add/remove of these specs links them to their
  trigger file.
- New magic xattr `bundle_trigger` (`1`/`0`) indicating whether a file is a
  bundle trigger.

Tested with `test/src/709-cvmfsbundles` and unit tests for the bundle and
bundle-file managers.

For more details, see also [this contribution at CHEP 2026](https://indico.cern.ch/event/1471803/contributions/6967096/).

Assisted-by: Claude (claude-opus-4-8)

vvolkl committed:
add comment on file-bundle

vvolkl committed:
add comment on file-bundle

jblomer committed:
Add publisher support to add and remove file bundle specs (#4001)

File bundle specifications are linked to a bundle trigger file in the
same directory.

---------

Co-authored-by: Valentin Volkl <valentin.volkl@cern.ch>

jblomer committed:
Add publisher support to add and remove file bundle specs (#4001)

File bundle specifications are linked to a bundle trigger file in the
same directory.

---------

Co-authored-by: Valentin Volkl <valentin.volkl@cern.ch>

christge-sft committed:
Create the mechanism to retrieve the files from the json bundle

christge-sft committed:
Create the mechanism to retrieve the files from the json bundle

christge-sft committed:
Spawn a dedicated thread to fetch a file within a bundle

christge-sft committed:
Spawn a dedicated thread to fetch a file within a bundle

christge-sft committed:
Integrate the BundleMgr to cvmfs_open()

christge-sft committed:
Integrate the BundleMgr to cvmfs_open()

christge-sft committed:
Make feature runtime selectible

christge-sft committed:
Make feature runtime selectible

DrDaveD commented on cvmfs/cvmfs.cc:
```diff
@@ -1179,11 +1179,19 @@ static void cvmfs_open(fuse_req_t req, fuse_ino_t ino,
-  if (dirent.IsBundleTrigger()) {
+  std::string is_prefetching_enabled;
+  if (options_mgr_) {
+    options_mgr_->GetValue("CVMFS_FUSE_NOTIFY_INVALIDATION",
```
I'm confused by this name.  What does that have to do with prefetching?

vvolkl replied:
Ys, that's a copy-paste artefact. I have some updates to this branch that my presentation is based on, will push soon.

vvolkl committed:
thread pool, parallel decompression, and remaining fixes for chep26

vvolkl committed:
small fixes

vvolkl committed:
Merge branch 'devel' into filebundle

christge-sft committed:
Suggestions on the pool of fetchers version (#4246)

1. I change the pool of fetchers from a `std::vector<pthread_t>*` to a
`std::vector<std::unique_ptr<pthread_t>>`
The motivation here is that there is no need to keep the pool vector via
a pointer (minor)


2. I changed the pool creation policy.
In the old version first we'ld default construct threads for the entire
pool, and in case of an initialization failure we'ld default destruct
what was not initialized properly.
Now I just increment the pool with the properly initialized fetcher
threads.

+ Extra:
Determines if feature is used during `Init()`. This should make the
decision on whether we should look for a trigger file in the first place
faster (important when feature is disabled).

vvolkl committed:
Merge branch 'devel' into filebundle

vvolkl committed:
fix typo

vvolkl committed:
fix compilation with c++11

vvolkl committed:
more c++11 compatibility

christge-sft committed:
Cleanup Code. (#4273)

Remove TODO on fetching- the while loop is eventually well justified.
Run clang-format.

christge-sft committed:
Create an integration test for file bundle prefetching. (#4279)
